driver-core/src/main/com/mongodb/client/model/expressions/Expression.java

```diff
@@ -21,104 +21,310 @@
- * be thought of as boxed values. Evaluation of an expression will usually occur
- * on a MongoDB server.
+ * A value in the context of the MongoDB Query Language (MQL).
  *
- * <p>Users should treat these interfaces as sealed, and must not implement any
- * expression interfaces.
+ * <p>The API provided by this base type and its subtypes is the Java-native
+ * variant of MQL. It is used to query the MongoDB server, to perform remote
+ * computations, to store and retrieve data, or to otherwise work with data on
+ * a MongoDB server or compatible execution context. Though the methods exposed
+ * through this API generally correspond to MQL operations, this correspondence
+ * is not exact.
  *
- * <p>Expressions are typed. It is possible to execute expressions against data
- * that is of the wrong type, such as by applying the "not" boolean expression
- * to a document field that is an integer, null, or missing. This API does not
- * define the output in such cases (though the output may be defined within the
- * execution context - the server - where the expression is evaluated). Users of
- * this API must mitigate any risk of applying an expression to some type where
- * resulting behaviour is not defined by this API (for example, by checking for
- * null, by ensuring that field types are correctly specified). Likewise, unless
- * otherwise specified, this API does not define the order of evaluation for all
- * arguments, and whether all arguments to some expression will be evaluated.
+ * <p>The following is an example of usage within an aggregation pipeline. Here,
+ * the current document value is obtained and its "numberArray" field is
+ * filtered and summed, in a style similar to that of the Java Stream API:
+ *
+ * <pre>{@code
+ * import static com.mongodb.client.model.expressions.Expressions.current;
+ * MongoCollection<Document> col = ...;
+ * AggregateIterable<Document> result = col.aggregate(Arrays.asList(
+ *     addFields(new Field<>("result", current()
+ *         .<MqlNumber>getArray("numberArray")
+ *         .filter(v -> v.gt(of(0)))
+ *         .sum(v -> v)))));
+ * }</pre>
+ *
+ * <p>Values are typically initially obtained via the current document and its
+ * fields, or specified via statically-imported methods on the
+ * {@link Expressions} class.
+ *
+ * <p>As with the Java Stream API's terminal operations, corresponding Java
+ * values are not directly available, but must be obtained indirectly via
+ * {@code MongoCollection.aggregate} or {@code MongoCollection.find}.
+ * Certain methods may cause an error, which will be produced
+ * through these "terminal operations".
+ *
+ * <p>The null value is not part of, and cannot be used as if it were part
+ * of, any explicit type (except the root type {@link Expression} itself).
+ * See {@link Expressions#ofNull} for more details.
+ *
+ * <p>There is no explicit "missing" or "undefined" value. Users may use
+ * {@link MapExpression#has}.
```

Comment:
We discussed that the intent here was to express that the only special value that is not of any explicit type other than `Expression` is the null value, which can be expressed via `Expressions.ofNull`.

The current wording _"no explicit "missing" or "undefined" value"_ implies (at least that's my interpretation) that there are such implicit values (can't be expressed, unlike null, but are expected to be encountered). Maybe instead of the two highlighted sections, we could just say something like

> `{@linkplain Expressions#ofNull Null}` is a special value that is of the `{@link Expression}` type and is not of any other explicit type. The API does not specify other special values.

If not, let's still somehow make it clear that neither "missing", nor "undefined", nor "any other special value" is specified.

Reply:
I think the "cannot be used as if it were" wording/emphasis is important for Java developers. I think calling it "special" might be confusing. I think we still need to call out missing/undefined specifically, since they are mentioned in the server docs. I have updated the wording.